lib/puppet/util/autoload.rb

```diff
@@ -55,10 +55,12 @@ def changed?(name)
     # Load a single plugin by name.  We use 'load' here so we can reload a
     # given plugin.
     def load_file(name, env)
+      load_path = $LOAD_PATH
       file = get_file(name.to_s, env)
       return false unless file
       begin
         mark_loaded(name, file)
+        $LOAD_PATH.push *(search_directories(env))
```

Comment:
`search_directories` will include everything already on the `$LOAD_PATH` since its definition is `[gem_directories, module_directories(env), libdirs(), $LOAD_PATH].flatten`. I think the result is lots of extra entries on the path. This has performance implications as it will search through all of them.